File: library/src/main/java/com/coolerfall/download/OkHttpDownloader.java
package com.coolerfall.download;

import android.net.Uri;

import java.io.IOException;
import java.io.InputStream;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okio.Buffer;
import okio.BufferedSource;

import static com.coolerfall.download.Utils.CONTENT_DISPOSITION;
import static com.coolerfall.download.Utils.DEFAULT_CONNECT_TIMEOUT;
import static com.coolerfall.download.Utils.DEFAULT_READ_TIMEOUT;
import static com.coolerfall.download.Utils.DEFAULT_WRITE_TIMEOUT;
import static com.coolerfall.download.Utils.HTTP;
import static com.coolerfall.download.Utils.HTTPS;
import static com.coolerfall.download.Utils.HTTP_TEMP_REDIRECT;
import static com.coolerfall.download.Utils.LOCATION;
import static com.coolerfall.download.Utils.MAX_REDIRECTION;
import static com.coolerfall.download.Utils.getFilenameFromHeader;
import static java.util.concurrent.TimeUnit.MILLISECONDS;

/**
 * A downloader implemented by {@link OkHttpClient}.
 *
 * @author Vincent Cheung (coolingfall@gmail.com)
 */
public final class OkHttpDownloader implements Downloader {
	private final OkHttpClient client;
	private Response response;
	private int redirectionCount = 0;

	private static OkHttpClient defaultOkHttpClient() {
		return new OkHttpClient.Builder().connectTimeout(DEFAULT_CONNECT_TIMEOUT, MILLISECONDS)
			.readTimeout(DEFAULT_READ_TIMEOUT, MILLISECONDS)
			.writeTimeout(DEFAULT_WRITE_TIMEOUT, MILLISECONDS)
			.build();
	}

	/**
	 * Create an instance using a default {@link OkHttpClient}.
	 *
	 * @return {@link OkHttpDownloader}
	 */
	public static OkHttpDownloader create() {
		return new OkHttpDownloader(null);
	}

	/**
	 * Create an instance using a {@code client}.
	 *
	 * @return {@link OkHttpDownloader}
	 */
	public static OkHttpDownloader create(OkHttpClient client) {
		return new OkHttpDownloader(client);
	}

	private OkHttpDownloader(OkHttpClient client) {
		this.client = client == null ? defaultOkHttpClient() : client;
	}

	@Override public String detectFilename(Uri uri) throws IOException {
		redirectionCount = 0;
		Response response = innerRequest(client, uri, 0);
		String url = response.request().url().toString();
		String contentDisposition = response.header(CONTENT_DISPOSITION);
		response.close();
		return getFilenameFromHeader(url, contentDisposition);
	}

	@Override public int start(Uri uri, long breakpoint) throws IOException {
		redirectionCount = 0;
		response = innerRequest(client, uri, breakpoint);
		return response.code();
	}

	@Override public long contentLength() {
		return getContentLength(response.body());
	}

	@Override public InputStream byteStream() {
		return response == null ? null : response.body().byteStream();
	}

	@Override public void close() {
		if (response != null) {
			response.close();
		}
	}

	@Override public Downloader copy() {
		return create(client.newBuilder().build());
	}

	Response innerRequest(OkHttpClient client, Uri uri, long breakpoint) throws IOException {
		String scheme = uri.getScheme();
		if (!HTTP.equals(scheme) && !HTTPS.equals(scheme)) {
			throw new DownloadException(0, "url should start with http or https");
		}

		Request.Builder builder = new Request.Builder().url(uri.toString());
		if (breakpoint > 0) {
			builder.header("Accept-Encoding", "identity")
				.header("Range", "bytes=" + breakpoint + "-")
				.build();
		}
		Response response = client.newCall(builder.build()).execute();
		int statusCode = response.code();
		switch (statusCode) {
		case 301:
		case 302:
		case 303:
		case HTTP_TEMP_REDIRECT:
			response.close();
			if (redirectionCount++ < MAX_REDIRECTION) {
			    /* take redirect url and call start recursively */
				String redirectUrl = response.header(LOCATION);
				return innerRequest(client, Uri.parse(redirectUrl), breakpoint);
			} else {
				throw new DownloadException(statusCode, "redirects too many times");
			}
		}

		return response;
	}

	/* read response content length from server */
	long getContentLength(ResponseBody body) {
		if (body == null) {
			return -1;
		}

		long contentLength = body.contentLength();
		if (contentLength <= 0) {
			BufferedSource source = body.source();
			try {
				source.request(Long.MAX_VALUE);
				Buffer buffer = source.buffer();
				contentLength = buffer.size();
			} catch (IOException e) {
				return -1;
			}
		}

		return contentLength;
	}
}

File: library/src/main/kotlin/com/coolerfall/download/OkHttpDownloader.kt
package com.coolerfall.download

import okhttp3.OkHttpClient
import okhttp3.Response
import java.util.concurrent.atomic.AtomicInteger
import kotlin.Throws
import java.io.IOException
import android.net.Uri
import okhttp3.Request
import java.io.InputStream
import java.util.concurrent.TimeUnit.MILLISECONDS

/**
 * A downloader implemented by [OkHttpClient].
 *
 * @author Vincent Cheung (coolingfall@gmail.com)
 */
class OkHttpDownloader private constructor(client: OkHttpClient?) : Downloader {
  private val client: OkHttpClient = client ?: defaultOkHttpClient()
  private var response: Response? = null
  private val redirectionCount = AtomicInteger()

  companion object {

    private fun defaultOkHttpClient(): OkHttpClient {
      return OkHttpClient.Builder()
          .connectTimeout(Helper.DEFAULT_CONNECT_TIMEOUT.toLong(), MILLISECONDS)
          .readTimeout(Helper.DEFAULT_READ_TIMEOUT.toLong(), MILLISECONDS)
          .writeTimeout(Helper.DEFAULT_WRITE_TIMEOUT.toLong(), MILLISECONDS)
          .build()
    }

    /**
     * Create an instance using a default [OkHttpClient].
     *
     * @return [OkHttpDownloader]
     */
    @JvmStatic fun create(): OkHttpDownloader {
      return OkHttpDownloader(null)
    }

    /**
     * Create an instance using a `client`.
     *
     * @return [OkHttpDownloader]
     */
    @JvmStatic fun create(client: OkHttpClient?): OkHttpDownloader {
      return OkHttpDownloader(client)
    }
  }

  @Throws(IOException::class) override fun detectFilename(uri: Uri): String {
    redirectionCount.set(Helper.MAX_REDIRECTION)
    val response = innerRequest(client, uri, 0)
    val url = response.request.url.toString()
    val contentDisposition = response.header(Helper.CONTENT_DISPOSITION)
    response.close()
    return Helper.getFilenameFromHeader(url, contentDisposition)
  }

  @Throws(IOException::class) override fun start(
    uri: Uri,
    breakpoint: Long
  ): Int {
    redirectionCount.set(Helper.MAX_REDIRECTION)
    response = innerRequest(client, uri, breakpoint)
    return response!!.code
  }

  override fun contentLength(): Long {
    return response?.body?.contentLength() ?: -1
  }

  override fun byteStream(): InputStream? {
    return response?.body?.byteStream()
  }

  override fun close() {
    response?.close()
  }

  override fun copy(): Downloader {
    return create(client)
  }

  @Throws(IOException::class) fun innerRequest(
    client: OkHttpClient,
    uri: Uri,
    breakpoint: Long
  ): Response {
    val builder: Request.Builder = Request.Builder()
        .url(uri.toString())
    if (breakpoint > 0) {
      builder.header("Accept-Encoding", "identity")
          .header("Range", "bytes=$breakpoint-")
          .build()
    }
    val response = client.newCall(builder.build())
        .execute()
    when (val statusCode = response.code) {
      301, 302, 303, Helper.HTTP_TEMP_REDIRECT -> {
        response.close()
        return if (redirectionCount.decrementAndGet() >= 0) {
          /* take redirect url and call start recursively */
          val redirectUrl = response.header(Helper.LOCATION)
              ?: throw DownloadException(statusCode, "redirects got no `Location` header")
          innerRequest(client, Uri.parse(redirectUrl), breakpoint)
        } else {
          throw DownloadException(statusCode, "redirects too many times")
        }
      }
    }
    return response
  }
}